src/ray/raylet/node_manager.cc

```diff
@@ -837,33 +842,17 @@ void NodeManager::ProcessRegisterClientRequestMessage(
   auto message = flatbuffers::GetRoot<protocol::RegisterClientRequest>(message_data);
   auto client_id = from_flatbuf<ClientID>(*message->worker_id());
   client->SetClientID(client_id);
-  auto worker = std::make_shared<Worker>(message->worker_pid(), message->language(),
-                                         message->port(), client);
+  Language language = static_cast<Language>(message->language());
+  auto worker = std::make_shared<Worker>(message->worker_pid(), language, message->port(),
+                                         client, client_call_manager_);
   if (message->is_worker()) {
     // Register the new worker.
-    bool use_push_task = (worker->Port() > 0);
-    if (use_push_task) {
-      auto entry = worker_clients_.find(client_id);
-      if (entry != worker_clients_.end()) {
-        RAY_LOG(WARNING) << "Received registration of a new worker that already exists: "
-                         << client_id;
-        return;
-      }
-
-      // Initialize a rpc client to the new worker.
-      std::unique_ptr<rpc::WorkerTaskClient> grpc_client(
-          new rpc::WorkerTaskClient("127.0.0.1", worker->Port(), client_call_manager_));
-      worker_clients_.emplace(client_id, std::move(grpc_client));
-    }
-
+    bool use_push_task = worker->UsePush();
```

Comment:
Nit, please move this to be above the `if (use_push_task)` statement that uses it below, and move the `auto connection ...` line inside the if block.

Reply:
This is actually intentional as `worker_pool_.RegisterWorker` does a `std::move` and `worker` is no longer valid after that line.